src/lib.rs

```diff
@@ -54,6 +57,25 @@ pub fn rewrite(input: &str, mode: &Mode, force: bool) -> Result<String, TlaError
     //println!("{:#?}", tla_lines);
     replace_symbols(&mut tla_lines);
 
+    // push a empty TlaLine,
+    // tla_lines...join("\n") will add a '\n' into end of the output,
+    // the reason why we don't push '\n' into output string directly is
+    // that maybe cause a reallocation.
+    if end_of_newline {
```

Comment:
Move this functionality into the `Tla_lines::construct_from` method

Reply:
> Move this functionality into the `Tla_lines::construct_from` method

After re-thinking about that, I think changing `tla_lines` for output is bad idea, if we put the functionality into `Tla_lines::construct_from`, that will be very confused, for examples, the input `"abc\n"` is exact one line by definition, so `Tla_lines::construct_from` should return `["abc"]` instead of `["abc", ""]`.

The newline problem is caused by the output logic, so I write the `TlaLine::output_from_lines` to handle that.